README
Note about current development
==============================
Subconvert is currently developed on ```gui``` branch (where heavy refactoring
is ongoing)! It's considered really unstable (and probably unusable), but if 
you're into the latest news, please take a look.

Subconvert - movie subtitles converter.
====================

Table of contents:
0 Capabilities
1 User manual
    1.1 Installation
        1.1.1 Uninstalling
        1.1.2 Reinstalling
        1.1.3 Dependencies
    1.2 Usage
        1.2.1 Available options
        1.2.2 Available formats
2 Programmer manual
3 License

====================

[ 0 Capabilities ]
    * fast and lightweight
    * easy to use: at minimum select subtitle file and you're done.
    * available as a window (GUI) and terminal (CLI) application
    * supports multiple subtitle formats
    * automatically checks video files for FPS information
    * converts between frame and time formats
    * supports as many file encodings as Python standard library does
    * detects encodings automatically
    * translatable (via gettext)

[ 1 User manual ]
This section describes installation and usage of subconvert.

[ 1.1 Installation ]
Since a version 0.8.3, a new Linux installer has been added. It is located in
the the main Subconvert directory and you can run it as follows:
    $ ./install.sh install

The installer will then ask you about install path. It should be a directory
where 'bin', 'share', 'share/doc' and 'share/locale' subdirectories reside.
Usually it will be simply '/usr' directory (you can choose it by pressing enter
on your keyboard on that step of installation), but you can also choose another
directory, like your home. Subconvert will work fine as long as
'chosen_path/bin' will be in your $PATH. It may not be visible on your system
menu though because most desktop environments search for application information
in /usr/share/applications.

You can specify install directory in script invocation (note the second
argument in a command below):
    $ ./install.sh install user/specified/path

Remember that if you want to install Subconvert outside of your HOME directory,
you'll need root privileges.

[ 1.1.2 Uninstalling ]
Install script allows you to remove Subconvert the same way, with optional path
argument:
    $ ./install.sh uninstall [user/specified/path]

At the moment the script will REQUIRE specifying earlier or later install path
(or -- being more specific -- PREFIX). This might change in next releases. If
you don't remember where Subconvert is installed, you can test it with common
Unix command, 'which':
    $ which subconvert

This will point you to a symbolic link in a ${PREFIX}/bin. ${PREFIX} is a
directory that you're looking for.

[ 1.1.3 Reinstalling ]
Reinstalling is very similar to installing and reinstalling. In fact, install
script simply executes those two actions one after another:
    $ ./install.sh reinstall [user/specified/path]

[ 1.1.4 Dependencies ]
    + Python 2.6+
    + python-qt4
    + python-chardet (optional, recommended)
    + MPlayer (optional, recommended)

[ 1.2 Usage ]
    subconvert [options] input_file [input_file...]

You can specify as many input files as you like to begin batch converting them.
You can also use common Unix wildcards:
    subconvert [options] *
    subconvert [options] *.txt
    subconvert [options] common_prefix*.srt
    subconvert [options] subtitle.s{rt,ub}

[ 1.2.1 Available options ]
Usage: subconvert.py [options] input_file [input_file...]

Options:

  General options:

    --version
        Show program's version number and exit.

    -h, --help
        Show this help message and exit.

    -f, --force
        Force all operations without asking (assuming yes).

    -q, --quiet
        Silence Subconvert output.

    --debug
        Generate debug output.

  File options:
    Options that describe how subtitle files should be handled.

    -e ENCODING, --encoding=ENCODING
        Input file encoding. If no encoding is provided, Subconvert will try to
        automatically detect file encoding and switches to 'UTF-8' when
        unsuccessfull. For a list of available encodings, see:
        http://docs.python.org/library/codecs.html#standard-encodings.

    --output-encoding=OUTPUT_ENCODING
        Output file encoding. If no output encoding is provided, subconvert
        will save output files with the same encoding as input.

    -t FORMAT_, --format=FORMAT_
        Output subtitle format. Default: subrip. See [ 1.2.2 Available formats ]
        for a list of available formats

    -x EXT, --extension=EXT
        Specify extension of the output file if the default one is not what you
        like.

  Movie options:
    Options that describe the corelations between subtitle and movie files.

    --fps=FPS
        Movie/subtitles frames per second. Default: 25.

    -A, --auto-fps
        Automatically try to get fps from MPlayer.

    -v MOVIE_FILE, --video=MOVIE_FILE
        Movie file to get FPS info from.

EXAMPLES:
    subconvert subtitles.txt
        Create subtitles.srt, UTF-8 encoded (subrip wit srt extension is default
        format and UTF-8 is encoding fallback)
    subconvert *.txt
        Convert all txt files in a directory
    subconvert -e cp1250 subtitles.txt
        Convert subtitles.txt to subrip. Leave original file encoding which is
        explicitly set to cp1250.
    subconvert -t microdvd --fps=23.976 subtitles.srt
        Create subtitles.sub microdvd format. Conversion rate from time format
        (subrip) to frame format (microdvd) is 23.976 FPS.

[ 1.2.2 Available formats ]
    -t FORMAT and --format=FORMAT options indicate the subtitle format that
    output should be written to. Currently available subtitle formats are:

    * microdvd
    * subrip
    * tmp
    * subviewer
    * mpl2

[ 2 Programmer manual ]
The main application uses GenericSubParser class and its subclasses which are
automatically loaded. Each of those subclasses should overwrite some attributes
and methods which are called from the GenericSubParser.parse() method (that one
MUST NOT be overwritten as it's the heart and the brain of the application).
This method works as a generator and parsing results are yielded for each
subtitle section. Parsing results are given as a dictionary:
    { sub_no, sub_fmt*, sub: { time_from**, time_to**, text } }
    * 'time' or 'frame'
    ** FrameTime format

You should overwrite the following constants:

__SUB_TYPE__
    Name of subtitle format.

__OPT__
    String which represents a format name that a user can select (-t and
    --format options). For example, if user writes subconvert -t your_opt, then
    a subclass with __OPT__ == 'your_opt' will be chosen to handle subtitle
    converting and formatting.

__EXT__
    Extension of the output file used if user haven't specified it through the
    options.

__FMT__
    Specify subtitle type (either 'frame' or 'time'). You don't have to
    overwrite it -- if it's not specified, parser will try to determine it
    automatically. Keep in mind though that it's only a simple regex and doesn't
    have to be accurate every time, so it's better to write it explicitly.

__WITH_HEADER__
    Set to True if subtitle format contains some kind of header.

__MAX_HEADER_LEN__
    Performance option. Maximum line number to which parser will check for
    header. By default it is set to 50 but you can change it if you want.

end_pattern
    This pattern is searched in every line of input file. When it's matched, it
    means that a single subtitle has been found. Further searches (described
    below) are going to be made on that subtitle and if they are successfull, a
    parsing result is yielded and sub_section string is cleared.
    For example, for a subrip file it might be a double endline. Only when we
    found a single subtitle section, we're able to get from it time, formatting,
    etc.

pattern
    Regex pattern that operates on a found subtitle section. It should match
    groups named <text>, <time_from> and possibly <time_to>, where <text> is a
    subtitle text and <time_from> and <time_to> are subtitle timestamps or
    frames (depending on subtitle format).

sub_fmt
    Subtitle format string. It indicates how a subtitle block should look like.
    It contains of tags that are understandable to Subconvert. This format
    string is used place a specific part of parsed subtitle into a correct place
    in new one.
    Allowed tags: {gsp_no}, {gsp_from}, {gsp_to}, {gsp_text}

sub_formatting
    A dictionary containing subtitle text formatting tags. Those tags indicate
    formatting which some subtitle formats allow (or, telling the truth, some
    movie players allow, e.g. not all players support subrip text formatting).
    There are both opening and closing tags to indicate formatted parts of text.
    When a subtitle is converted to the specific format, those tags are changed
    to the correct formatting for that format according to the specifications
    included in sub_formatting dictionary.
    Allowed tags: {gsp_x_} for opening tags and {_gsp_x} for closing tags,
    where x is one of the following: b, i, u (for bold, italics and underline)
    Also, {gsp_nl} is allowed to indicate line breaks.
    Example: sub_formatting = {'gsp_nl': r'|',}

There are also a couple of methods that usually should be overwritten:

def __init__(self, filename, fps, encoding, lines=None)
    Don't forget to call GenericSubParser.__init__(args) from your subclass
    constructor.

def get_header(self, section, atom)
    Called during parsing when input format contains header (__WITH_HEADER__ is
    set to True). This function should try to find and parse header in a given
    sub_section and save the results as a dictionary to atom['header']. If
    parsing was successfull it should return True and False otherwise. Note that
    though programmer is not required to follow any naming convention, it is
    HIGHLY RECOMMENDED to name atom['header'] keys with full, lower case,
    underscore separated names (like 'information', 'delay', 'author',
    'version', 'title' etc.). It is to allow sharing those pieces of
    informations between various formats.
    For formats with __WITH_HEADER__ set to True this method is checked until it
    returns True (or __MAX_HEADER_LEN__ is reached). After that it is not called
    again

def convert_header(self, header)
    Converts a given header (which is dictionary saved by get_header method) to
    a string that can be saved to a file. It is HIGHLY RECOMMENDED to define
    default values for all required pieces of informations as subtitle formats
    do not have to share the same informations between each other.

def format_text(self, string)
    Converts string formatting in a given string from a subtitle-format specific
    to the one recognised by GenericSubParser.  Specific subtitle formatting
    should be changed with tags described in sub_formatting constant. It
    can be done any way you prefer (regex, string manipulation, ...).
    Important note: don't forget to escape any curly braces ('{' and '}') that
    may occur in subtitle text as they might break things down.

def str_to_frametime(self, string)
    Converts strings which are recognised as a timestamp to FrameTime objects.
    It is important to have a common time format which can be passed, recognised
    and operated on by various GenericSubParser subclasses (subtitle formats).

def get_time(self, ft, which)
    Opposite operation to 'str_to_frametime'. Extracts time (time_from and
    time_to) from a given FrameTime object to properly formatted string which
    represents a time format specific for an output subtitle format. There is
    also a possibility to specify different formats for 'time_from' and
    'time_to' thanks to 'which' argument.

Please note that if you feel ok with defaults of all above constants attributes,
you don't have to even touch them. Read the code for more info!

[ 3 License ]
Subconvert is free software licensed under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
For details, see <LICENSE.txt> or http://www.gnu.org/licenses/gpl.html


CHANGELOG
Version history

1.0.3   : 19.04.2013 (contains not released fixes from 1.0.2)
* ADD:  Enabled menu bar subtitles editing (+undo/redo)
* ADD:  "Save as..." opens a file in new tab to which subs are saved.
* ADD:  Inserting and removing subtitles.
* ADD:  Jumping to the start of highlighted sub
* FIX:  Fixed not showing errors on "save all"
* FIX:  Show a message box when there's an error on SPF saving
* FIX:  Update "subtitles menu" on file list selection change
* FIX:  Strip subtitles from whitespaces
* FIX:  Fixed loading autoInputEncoding SPF flag
* FIX:  Fixed displaying timestamp for some kind of movies
* FIX:  Added missing window icon
* MOD:  Change SubtitleProperties format: pickle -> JSON
* MOD:  Changed forward/rewind shortucts to ctrl+right, ctrl+left

1.0.1   : 21.12.2013
* ADD:  spinbox on subtitle time edit
* FIX:  installing Subconvert icons
* FIX:  missing icons
* MOD:  FPS value from pfile will be used when there's no FPS specified

1.0.0   : 10.12.2013 (Brave New World)
* ADD:  complete redesign of GUI which has a lot of cool features:
        * subtitle editing
        * undo/redo
        * batch or single file works
        * reencoding of subtitle file
        * side panel
* ADD:  introduced Subtitle Property Files
* ADD:  introduced output file syntax ("%f" substitution)
* FIX:  correct internationalization (moved *.po files + use of ngettext)
* MOD:  HUGE code refactoring
* MOD:  changed hand made install script to autotools
* MOD:  migrate Subconvert to Python 3
* MOD:  subconvert by default spawns GUI. Use '-c' to access commandline
        interface

0.9     : 12.05.2012 (Code Injector)
* ADD:  "-g" option to SubConvert CLI version
* ADD:  #32: added icon for the GUI file list which indicates a status of a job
        for that file
* FIX:  #42: MPL2 error with one-digit time indicators
* FIX:  A lot of (hopefully all) filename encoding issues are now fixed
* FIX:  Two untranslated strings are now translatable
* FIX:  SubOptParser gettext issues
* FIX:  Help text unification
* FIX:  #35: silence MPlayer error
* FIX:  #33: omit Byte Order Mark in UTF8 files
* FIX:  proper arguments parsing in GUI
* MOD:  Rewritten command line options
* MOD:  #40: GUI: change the way that logging is handled. Most significant
        change are individual logs for each file on the list (available after
        double-click on that file)

0.8.3   : 11.02.2012 (Talkative Hiker)
* ADD:  Polish translation
* ADD:  Auto detection of more movie formats (both GUI and CLI)
* ADD:  New install script
* FIX:  SubConvert detects both upper case and lower case movie extensions
* FIX:  with '-v' option, SubConvert will not require any additional parameters
        and will simply try to get FPS from a given file
* FIX:  #25: changed string concat method in MicroDVD
* FIX:  Tests work again
* FIX:  Crash when input string contained non-ascii characters
* FIX:  SubConvert will now try to auto detect locale path
* MOD:  'Auto FPS' checkbox is now checked on GUI start
* RMV:  Combo box from GUI with an output sub extension

0.8.2   : 05.01.2012 (Working Student)
* ADD:  Unit tests
* ADD:  option to specify output file(s) encoding
* ADD:  MPL2 format support
* FIX:  HUGE code refactoring
* FIX:  ugly converting error (which doubled first subtitle and removed the
        last one)
* FIX:  incorrect SubViewer miliseconds converting
* FIX:  SubRip skipping last sub when no new line at the EOF (now Generic
        Parser forces parsing a section at the EOF)
* FIX:  smart skipping of incorrect empty lines
* RMV:  custom updater

0.8.1   : 07.12.2011
* ADD:  SubConvert updater
* ADD:  encoding detecting (optional)

0.8.0-1 : 04.12.2011
* FIX:  unhandled assertion errors which were caused by incorrect subtitles

0.8     : 28.11.2011
* ADD:  setup.py (installer)
* ADD:  GUI
* ADD:  option to stop converting when file exists. Not only it stops file from
        being updated but also quits batch work.
* FIX:  remove file extension from subviewer's [TITLE]
* FIX:  let the shell find python in $PATH
* FIX:  Add maximum file size limit to subtitle files.
* FIX:  Better parsing state output.
* FIX:  Don't save not completely parsed files anymore.

0.7     : 19.11.2011
* ADD:  TMP sub formatting
* ADD:  SubViewer sub formatting
* ADD:  full converting subtitles without sub end time
* ADD:  support subtitles with header
* FIX:  error on -m option mistyped
* FIX:  error that prevented sub parsing when no sub end time was present

0.6.2   : 13.11.2011
* ADD:  mplayer auto-check more movie formats: mkv, mpg, mp4 and wmv
* ADD:  handle multiple input files
* ADD:  option to specify output file(s) extension
* FIX:  SubRip incorrect parsing (trivia: in my test subtitles a word which
        starts with 'f' appears for the first time in 119th subtitle)
* FIX:  not clear notifications about parsing and IO operations state + add some
        more info to verbose output

0.6.1   : 10.11.2011
* FIX:  Uniform backup system
* FIX:  A lot of code refactoring. Now works slightly faster

0.6     : 10.11.2011
* ADD:  getting FPS from avi files using MPlayer
* ADD:  option to backup files if they exist (overwriting input files are back
        them up automatically while output files are backed up explicitly)
* FIX:  code refactoring. Until now subconvert reads whole files to the memory
        which speeds it up almost twice
* FIX:  bug where parsing empty files returned results

0.5.2   : 09.11.2011
* ADD:  force option
* ADD:  verbose option
* FIX:  writing a newly created file
* FIX:  unescaped curly braces in microdvd and subrip that caused crashes

0.5     : 09.11.2011
* ADD:  SubRip and MicroDVD parsers/handlers
* ADD:  Writing output to a file
* ADD:  Automatic handling all newly implemented file formats


install.sh
#!/bin/bash

DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

function printUsage {
    echo "Usage:"
    echo "   install.sh install| uninstall | reinstall | depcheck [install_path]"
}

# Args 3 and 4 are passed through reference:
# $1 - import name; $2 - module nape (installed through e.g. apt, Pacman, etc.)
# $3 - ${DEPRET}, $4 - ${DEPNEED} to update
function check_python_modules {
    if [[ ${1} && ${2} && ${3} ]]; then
        if [[ ! -e subconvertpycheck.py ]]; then
            touch subconvertpycheck.py
        fi

        echo "import ${1}" > subconvertpycheck.py
        python subconvertpycheck.py 2> /dev/null

        if [[ $? -ne 0 ]]; then
            echo "[ERR] ${1} not found."
            DEPRET=1
            DEPNEED=$(echo ${4} ${2})
        else
            echo "[INF] ${1} found."
        fi

        rm -f subconvertpycheck.py
    else
        echo "[!!!] Incorrect arguments in check_python_modules! This is installer bug and should be reported!"
        echo "[!!!] Debug output: ${1} ;; ${2} ;; ${3} ;; ${4}"
        exit 1
    fi
}

# Args:
# command, printable name, error level, ${DEPRET}, ${DEPNEED}
function check_software {
    if [[ ${1} && ${2} && ${3} && ${4} ]]; then
        ${1} > /dev/null 2>&1
        if [[ $? -ne 0 ]]; then
            echo "${3} ${2} not found."
            if [[ ${3} == "[ERR]" ]]; then
                DEPRET=1
            fi
            DEPNAME=$(echo ${1} | sed "s/ .*//")
            DEPNEED="${5} ${DEPNAME}"
        else
            echo "[INF] ${2} found."
        fi
    else
        echo "[!!!] Incorrect arguments in check_software! This is installer bug and should be reported!"
        echo "[!!!] Debug output: ${1} ;; ${2} ;; ${3} ;; ${4} ;; ${5}"
        exit 1
    fi

}

function check_dependencies {
    # dependencies: python2.6, python-qt4, python-chardet, mplayer, gettext
    DEPRET=0
    DEPNEED=""

    PYTHON_VERSION=$(python --version 2>&1 | sed "s/Python //") # Might be 2.7.2+
    PYTHON_BIG_VERSION=$(echo ${PYTHON_VERSION} | grep -o "^[0-9]\+.[0-9]\+" | sed "s/\.//")

    # Check Python version
    if [[ PYTHON_BIG_VERSION -lt 26 ]]; then
        echo "[ERR] Python version: ${PYTHON_VERSION}. Required: >2.6"
        DEPRET=1
        DEPNEED="${DEPNEED} python2.6"
    else
        echo "[INF] Python version: ${PYTHON_VERSION}."
    fi


    # Python libs
    check_python_modules "PyQt4" "python-qt4" ${DEPRET} ${DEPNEED}
    check_python_modules "chardet" "python-chardet" ${DEPRET} ${DEPNEED}

    # Software
    check_software "gettext --version" "gettext" "[WRN]" ${DEPRET} ${DEPNEED}
    # mplayer -v is not a proper mplayer switch but it doesn't produce an error
    check_software "mplayer -v" "MPlayer" "[ERR]" ${DEPRET} ${DEPNEED}


    if [[ ${DEPNEED} != "" ]]; then
        echo -ne "------------------------------\nNot all dependencies are installed on the system. "
        if [[ ${DEPRET} -eq 0 ]]; then
            echo -n "Note that some of them are not required (see manual for reference). "
        fi
        echo -e "You might consider copying the following list of not installed packages to your package manager:\n"
        echo -e "    ${DEPNEED}\n------------------------------"
    fi

    return ${DEPRET}
}

function remove {
    if [[ -e "${1}" ]]; then
        if [[ $# -eq 2 && ${2} == "dir" ]]; then
            echo "[INF] Removing directory ${1}..."
            rm -rf "${1}"
        else
            echo "[INF] Removing file ${1}..."
            rm -f "${1}"
        fi
    else
        echo "[WRN] ${1} doesn't exist!"
    fi
}

function getPath {
    # ${1} provides a "prompt" text in this function
    if [[ $# -eq 0 ]]; then
        echo "[ERR] I cowardly refuse to do nothing..."
        exit 1
    else
        DEFAULT_DIR=/usr
        if [[ "${INSTALL_DIR}" == "" ]]; then
            echo "${1} [Default (press enter to use): ${DEFAULT_DIR}]:"
            echo -n "> "
            read INSTALL_DIR
        fi
        if [[ "${INSTALL_DIR}" == "" ]]; then
            INSTALL_DIR=${DEFAULT_DIR}
        else
            INSTALL_DIR="${INSTALL_DIR}"
        fi

        SUBC_BIN="${INSTALL_DIR}/bin"
        SUBC_SHARE="${INSTALL_DIR}/share/subconvert"
        SUBC_LOCALE="${INSTALL_DIR}/share/locale"
        SUBC_DOC="${INSTALL_DIR}/share/doc/subconvert"
    fi
}

function installLocales {
    for LDIR in ${DIR}/subconvert/locale/*; do
        POFILE="${LDIR}/LC_MESSAGES/subconvert.po"
        MOFILE="${LDIR}/LC_MESSAGES/subconvert.mo"
        if [[ -e ${POFILE} ]]; then
            if [[ -e ${MOFILE} ]]; then
                echo "[INF] Removing existing ${MOFILE}"
                rm -f ${MOFILE}
            fi
            msgfmt ${POFILE} -o ${MOFILE}
            MODEST="${SUBC_LOCALE}/$(basename ${LDIR})/LC_MESSAGES/"
            cp ${MOFILE} ${MODEST} 2> /dev/null
            if [[ $? -ne 0 ]]; then
                mkdir -p ${MODEST}
                cp ${MOFILE} ${MODEST}
            fi
        fi
    done
}

function installSubConvert {
    if [[ ! -e ${INSTALL_DIR} ]]; then
        echo -e "[ERR] It seems that '${INSTALL_DIR}' doesn't exist. Keep in mind that you have to provide path to existing directory in which a 'subconvert' directory can be created. Frequent choices are user HOME directory or '/usr' directory."
        echo "Please try to run the install script again."
        exit 1
    fi

    if [[ -e ${SUBC_SHARE} || -e ${SUBC_DOC} || -e ${SUBC_BIN}/subconvert || -e ${SUBC_BIN}/subconvert-gui ]]
    then
        echo "[ERR] It seems that previous version of SubConvert wasn't properly removed. Please try to uninstall it first."
        exit 1
    fi

    echo "[INF] Dependencies:"
    check_dependencies
    if [[ $? -ne 0 ]]; then
        echo "[ERR] Required packages are not installed."
        exit 1
    else
        echo "[INF] Dependency check passed."
    fi

    mkdir -p ${SUBC_SHARE}/subconvert
    if [[ $? -gt 0 ]]; then
        echo "[ERR] Cannot create ${SUBC_SHARE} directory. Do you have proper permissions?"
        exit 1
    fi

    echo "[INF] Copying files to ${SUBC_SHARE}"
    cp ${DIR}/{subconvert.py,subconvert-gui.py,install.sh} ${SUBC_SHARE} || exit 1
    cp ${DIR}/subconvert/*.py ${SUBC_SHARE}/subconvert
    cp -r ${DIR}/subconvert/subparser ${SUBC_SHARE}/subconvert
    cp -r ${DIR}/subconvert/subutils ${SUBC_SHARE}/subconvert
    cp -r ${DIR}/subconvert/img ${SUBC_SHARE}/subconvert
    chmod +x ${SUBC_SHARE}/{subconvert.py,subconvert-gui.py,install.sh}
    echo "[INF] Copying files to ${SUBC_DOC}..."
    if [[ ! -e ${SUBC_DOC} ]]; then
        echo "[INF] ${SUBC_DOC} doesn't exist. Creating it."
        mkdir -p ${SUBC_DOC}
    fi
    if [[ -d ${SUBC_DOC} ]]; then
        cp ${DIR}/{CHANGELOG,LICENSE.txt,README} ${SUBC_DOC}
    else
        echo "[WRN] ${SUBC_DOC} is not a directory! Skipping doc files!"
    fi
    echo "[INF] Creating symbolic links..."
    if [[ ! -e ${SUBC_BIN} ]]; then
        echo "[INF] ${SUBC_BIN} doesn't exist. Creating it."
        mkdir -p ${SUBC_BIN}
    fi
    ln -s ${SUBC_SHARE}/subconvert.py ${SUBC_BIN}/subconvert
    ln -s ${SUBC_SHARE}/subconvert-gui.py ${SUBC_BIN}/subconvert-gui

    if [[ ! -e ${INSTALL_DIR}/share/applications ]]; then
        echo "[INF] ${INSTALL_DIR}/share/applications doesn't exist. Creating it."
        mkdir -p ${INSTALL_DIR}/share/applications
    fi
    cp ${DIR}/subconvert.desktop ${INSTALL_DIR}/share/applications

    if [[ ! -e ${INSTALL_DIR}/share/icons/hicolor ]]; then
        echo "[WRN] ${INSTALL_DIR}/share/icons/hicolor doesn't exist."
        mkdir -p ${INSTALL_DIR}/share/icons/hicolor/{128x128,16x16,192x192,22x22,24x24,256x256,32x32,36x36,48x48,64x64,72x72,96x96,scalable}

        echo "[INF] Downloading hicolor theme."
        wget http://www.freedesktop.org/software/icon-theme/releases/hicolor-icon-theme-0.5.tar.gz
        if [[ $? -ne 0 ]]; then
            echo "[ERR] Installation of hicolor theme unsuccessfull! Stopping installation."
            exit 1
        fi

        echo "[INF] Installing hicolor theme."
        mkdir -p ${DIR}/hicolor_install
        mv hicolor-icon-theme-0.5.tar.gz ${DIR}/hicolor_install
        pushd .
        cd ${DIR}/hicolor_install
        tar -xzf hicolor-icon-theme-0.5.tar.gz
        cd hicolor-icon-theme-0.5
        cp index.theme ${INSTALL_DIR}/share/icons/hicolor/

        popd
        rm -rf ${DIR}/hicolor_install
    fi
    cp -r ${DIR}/subconvert/img/icons/* ${INSTALL_DIR}/share/icons/hicolor/

    if [[ ! -e ${INSTALL_DIR}/share/pixmaps ]]; then
        echo "[INF] ${INSTALL_DIR}/share/pixmaps doesn't exist. Creating it."
        mkdir -p ${INSTALL_DIR}/share/pixmaps
    fi
    ln -s ${INSTALL_DIR}/share/icons/hicolor/scalable/subconvert.svg ${INSTALL_DIR}/share/pixmaps/subconvert.svg
    ln -s ${INSTALL_DIR}/share/icons/hicolor/48x48/subconvert.png ${INSTALL_DIR}/share/pixmaps/subconvert-48.png
    ln -s ${INSTALL_DIR}/share/icons/hicolor/32x32/subconvert.png ${INSTALL_DIR}/share/pixmaps/subconvert-32.png

    echo "[INF] Creating and copying locales..."
    installLocales
    echo "Installation finished succesfully!"
}

function uninstallSubConvert {
    remove ${SUBC_BIN}/subconvert
    remove ${SUBC_BIN}/subconvert-gui
    remove ${SUBC_SHARE} "dir"
    remove ${SUBC_DOC} "dir"
    remove ${INSTALL_DIR}/share/pixmaps/subconvert.svg
    remove ${INSTALL_DIR}/share/pixmaps/subconvert-48.png
    remove ${INSTALL_DIR}/share/pixmaps/subconvert-32.png

    for SIZE in ${INSTALL_DIR}/share/icons/hicolor/*;
    do
        if [[ -d ${SIZE} && -e ${SIZE}/subconvert.png ]]; then
            remove ${SIZE}/subconvert.png
        elif [[ -d ${SIZE} && -e ${SIZE}/subconvert.svg ]]; then
            remove ${SIZE}/subconvert.svg
        fi
    done

    for LANG in ${INSTALL_DIR}/share/locale/*;
    do
        if [[ -d ${LANG} && -e ${LANG}/LC_MESSAGES/subconvert.mo ]]; then
            remove ${LANG}/LC_MESSAGES/subconvert.mo
        fi
    done
}

# Execution starts here

if [[ $# -lt 1 && $# -gt 2 ]]; then
    echo "Incorrect call of install script!"
    printUsage
    exit 1
fi

ACTION=${1}
INSTALL_DIR=${2}

case ${ACTION} in
    install)
        getPath "Select directory to install SubConvert"
        installSubConvert
        ;;
    uninstall)
        getPath "Select SubConvert PREFIX directory"
        uninstallSubConvert
        ;;
    reinstall)
        getPath "Reinstallation: Select SubConvert PREFIX directory"
        echo -e "\nTrying to uninstall..."
        uninstallSubConvert
        echo -e "\nInstalling new SubConvert..."
        installSubConvert
        ;;
    depcheck)
        echo "Scanning dependencies..."
        check_dependencies
        exit $?
        ;;
    *)
        echo "Incorrect call of install script!"
        printUsage
        exit 1
        ;;
esac

